ibis/impala/compiler.py

```diff
@@ -615,18 +615,49 @@ def _table_array_view(translator, expr):
 def _timestamp_op(func):
     def _formatter(translator, expr):
         op = expr.op()
-        arg, offset = op.args
-        formatted_arg = translator.translate(arg)
-        formatted_offset = translator.translate(offset)
+        left, right = op.args
+        formatted_left = translator.translate(left)
+        formatted_right = translator.translate(right)
+
+        if isinstance(left, (ir.TimestampScalar, ir.DateValue)):
+            formatted_left = 'cast({} as timestamp)'.format(formatted_left)
 
-        if isinstance(arg, ir.TimestampScalar):
-            formatted_arg = 'cast({} as timestamp)'.format(formatted_arg)
+        if isinstance(right, (ir.TimestampScalar, ir.DateValue)):
+            formatted_right = 'cast({} as timestamp)'.format(formatted_right)
 
-        return '{}({}, {})'.format(func, formatted_arg, formatted_offset)
+        return '{}({}, {})'.format(func, formatted_left, formatted_right)
 
     return _formatter
 
 
+_date_add = _timestamp_op('date_add')
+_timestamp_add = _timestamp_op('date_add')
+
+
+def _date_subtract(translator, expr):
+    op = expr.op()
+    left, right = op.args
+
+    if isinstance(right, ir.DateValue):
+        func = _timestamp_op('datediff')
+    else:
+        func = _timestamp_op('date_sub')
+
+    return func(translator, expr)
+
+
+def _timestamp_subtract(translator, expr):
+    op = expr.op()
+    left, right = op.args
+
+    if isinstance(right, ir.TimestampValue):
+        return 'unix_timestamp({}) - unix_timestamp({})'.format(
```

Comment:
Does this pass bigquery tests?

Reply:
`TimestampAdd`, `TimestampSubtract`, `DateAdd`, `DateSubtract` are disabled in bigquery, see [here](https://github.com/ibis-project/ibis/pull/1352/files#diff-57418510fe1d988e5cb1614da6247204R267)